### PR TITLE
roc_audio: adjust resampler coef 

### PR DESCRIPTION
**Pull request by dshil**
#85

### CODE REVIEW

File: src/lib/roc/config.h [hunk 1]
```diff
@@ -105,6 +105,9 @@ typedef struct roc_receiver_config {
 
     //! A bitmask of ROC_FLAG_* constants.
     unsigned int flags;
+
+    //! Number of samples for all channels per second.
+    unsigned int sample_rate;
```
Comment (gavv) [inline src/lib/roc/config.h]: This comment looks ambiguous. I suggest "Number of samples per second per channel". We should also fix corresponding comments in ReceiverConfig and SenderConfig.

File: src/modules/roc_audio/resampler_updater.cpp [hunk 2]
```diff
@@ -96,13 +98,15 @@ bool ResamplerUpdater::update(packet::timestamp_t time) {
         update_time_ += update_interval_;
     }
 
+    const float adjusted_coef = float(sample_rate_coef_ * fe_.freq_coeff());
+
     if (rate_limiter_.allow()) {
         roc_log(LogDebug, "resampler updater: queue_size=%lu fe=%.5f",
-                (unsigned long)queue_size, (double)fe_.freq_coeff());
+                (unsigned long)queue_size, (double)adjusted_coef);
```
Comment (gavv) [inline src/modules/roc_audio/resampler_updater.cpp]: It would be reasonable to print both freq_coeff and adjusted_coef here.

File: src/modules/roc_audio/resampler_updater.h [hunk 3]
```diff
@@ -35,7 +35,8 @@ class ResamplerUpdater : public packet::IWriter,
     //!  - @p update_interval defines how often to call FreqEstimator, in samples
     //!  - @p aim_queue_size defines FreqEstimator target queue size, in samples
     ResamplerUpdater(packet::timestamp_t update_interval,
-                     packet::timestamp_t aim_queue_size);
+                     packet::timestamp_t aim_queue_size,
+                     const float sample_rate_coef);
```
Comment (gavv) [inline src/modules/roc_audio/resampler_updater.h]: sample_rate_coef parameter should be documented.

File: src/modules/roc_pipeline/config.h [hunk 4]
```diff
@@ -135,7 +135,7 @@ struct ReceiverConfig {
     bool timing;
 
     ReceiverConfig()
-        : sample_rate(DefaultSampleRate)
+        : sample_rate(0)
```
Comment (gavv) [inline src/modules/roc_pipeline/config.h]: Why this change? The idea of default constructors of these *Config structures is that if a field was not overwritten explicitly, it has a reasonable value. This feature will be used in roc_lib, which will overwrite only those fields that were set to non-zero by user.

File: src/modules/roc_pipeline/receiver_session.cpp [hunk 5]
```diff
@@ -35,7 +36,8 @@ ReceiverSession::ReceiverSession(const SessionConfig& config,
 
     if (config.resampling) {
         resampler_updater_.reset(new (allocator_) audio::ResamplerUpdater(
-                                     config.fe_update_interval, config.latency),
+                                     config.fe_update_interval, config.latency,
+                                     (float)format->sample_rate / out_sample_rate),
```
Comment (gavv) [inline src/modules/roc_pipeline/receiver_session.cpp]: We should panic if out_sample_rate is zero. A panic is better than a segfault.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp [hunk 6]
```diff
@@ -71,13 +65,32 @@ bool Player::open(const char* name, const char* type) {
...
+            (unsigned long)output_->signal.channels);
+
     return true;
 }
 
 void Player::stop() {
     stop_ = 1;
 }
 
+void Player::start(pipeline::IReceiver* input) {
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Why a pointer? It's not allowed to be null, so a reference is preferred.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp [hunk 7]
```diff
@@ -71,13 +65,32 @@ bool Player::open(const char* name, const char* type) {
...
+    input_ = input;
+    core::Thread::start();
+}
+
+size_t Player::get_sample_rate() const {
+    if (output_) {
+        return size_t(output_->signal.rate);
+    }
+    return 0;
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: If output_ is null, it means that open() was not called or failed, and the user called get_sample_rate() anyway. It looks like a bug in the user code, so I think it's better to panic here instead of silently returning zero.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp [hunk 8]
```diff
@@ -115,7 +128,7 @@ void Player::loop_() {
     SOX_SAMPLE_LOCALS;
 
     while (!stop_) {
-        pipeline::IReceiver::Status status = input_.read(frame);
+        pipeline::IReceiver::Status status = input_->read(frame);
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: We should panic if run() was called and input_ is null. A panic is better than a segfault.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h [hunk 9]
```diff
@@ -60,11 +58,23 @@ class Player : public core::Thread {
...
+    //!  - @p input is used to read samples.
+    void start(pipeline::IReceiver* input);
+
     //! Stop thread.
     //! @remarks
     //!  Can be called from any thread.
     void stop();
 
+    //! Get sample rate
+    //!
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.h]: It's worth to mention how this sample rate is obtained.

Comment (gavv) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Also, we need a test for this feature.

File: src/tools/roc_recv/main.cpp [hunk 10]
```diff
@@ -148,6 +148,17 @@ int main(int argc, char** argv) {
...
+    sndio::Player player(sample_buffer_pool, allocator, args.oneshot_flag,
+                         config.channels, config.sample_rate);
+
+    if (!player.open(args.output_arg, args.type_arg)) {
+        roc_log(LogError, "can't open output file or device: %s %s", args.output_arg,
+                args.type_arg);
+        return 1;
+    }
+
+    config.sample_rate = player.get_sample_rate();
```
Comment (gavv) [inline src/tools/roc_recv/main.cpp]: I think we should check if sample_rate is zero and exit with error in this case.

File: src/lib/roc/config.h (hunk 1 above)
Comment (dshil) [reply src/lib/roc/config.h]: Fixed

File: src/modules/roc_audio/resampler_updater.h (hunk 3 above)
Comment (dshil) [reply src/modules/roc_audio/resampler_updater.h]: Fxd

File: src/modules/roc_audio/resampler_updater.cpp (hunk 2 above)
Comment (dshil) [reply src/modules/roc_audio/resampler_updater.cpp]: Fixed

File: src/modules/roc_pipeline/receiver_session.cpp (hunk 5 above)
Comment (dshil) [reply src/modules/roc_pipeline/receiver_session.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 7 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h (hunk 9 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 7 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/modules/roc_pipeline/receiver_session.cpp (hunk 5 above)
Comment (dshil) [reply src/modules/roc_pipeline/receiver_session.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 7 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/modules/roc_pipeline/receiver_session.cpp (hunk 5 above)
Comment (dshil) [reply src/modules/roc_pipeline/receiver_session.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 7 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h (hunk 9 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 8 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/tools/roc_recv/main.cpp (hunk 10 above)
Comment (dshil) [reply src/tools/roc_recv/main.cpp]: Fixed

File: src/modules/roc_pipeline/config.h (hunk 4 above)
Comment (dshil) [reply src/modules/roc_pipeline/config.h]: Providing zero sample rate for sox means, that it'll determine sample rate of the output file or device by its own.

UPD: Later I've checked, that if we provide unsupported sample rate to sox (e.g. 44K, while device supports 48K), sox will also detect the correct sample rate by its own.

I'll rollback this change.

Comment (dshil) [reply src/modules/roc_pipeline/config.h]: > This feature will be used in roc_lib, which will overwrite only those fields that were set to non-zero by user.

Where do we check that the user provides non-zero values?

```cpp
bool make_receiver_config(pipeline::ReceiverConfig& out, const roc_receiver_config* in) {
    out.default_session.latency = in->latency;
    out.default_session.timeout = in->timeout;
    out.default_session.samples_per_packet = in->samples_per_packet;
    out.sample_rate = in->sample_rate;
   
    //...
}
```

We'll panic only during a session creation.

**Comment by dshil** [comment]
ping @gavv

File: src/modules/roc_pipeline/config.h (hunk 4 above)
Comment (gavv) [reply src/modules/roc_pipeline/config.h]: > Later I've checked, that if we provide unsupported sample rate to sox (e.g. 44K, while device supports 48K), sox will also detect the correct sample rate by its own.

I think we should not rely on this, it's too implicit and can be changed when switching from SoX to something else.

I suggest the following:

* Use a reasonable default in ReceiverConfig::sample_rate (44100). This default will be ignored in roc_recv, but it will be used in roc_lib.

* Use a separate variable in roc_recv, initialize it with `--rate` option value or zero if it wan't given, pass it to SoX, then get the actual sample rate from SoX, and write it to ReceiverConfig.

> Where do we check that the user provides non-zero values?

There are no checks yet. However, I plan to add them in 0.1. It will look like:

```
if (in->sample_rate) {
    out.sample_rate = in->sample_rate;
}
```

File: src/modules/roc_audio/resampler_updater.cpp [hunk 11]
```diff
@@ -96,13 +98,16 @@ bool ResamplerUpdater::update(packet::timestamp_t time) {
         update_time_ += update_interval_;
     }
 
+    const float adjusted_coef = float(sample_rate_coef_ * fe_.freq_coeff());
```
Comment (gavv) [inline src/modules/roc_audio/resampler_updater.cpp]: Why do we need a cast?

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp [hunk 12]
```diff
@@ -71,18 +69,41 @@ bool Player::open(const char* name, const char* type) {
...
+        roc_panic("player: can't get sample rate for non-open output file or device");
+    }
+    return size_t(output_->signal.rate);
+}
+
 void Player::run() {
     roc_log(LogDebug, "player: starting thread");
 
+    if (!input_) {
+        roc_panic("player: thread is started before device or output file was opened");
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: The panic message is a bit misleading, because input_ is set in start(), not in open().

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h [hunk 13]
```diff
@@ -60,11 +58,23 @@ class Player : public core::Thread {
     //!  Should be called once before calling start().
     bool open(const char* name, const char* type = NULL);
 
+    //! Start reading samples.
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.h]: It's worth to mention that it starts a thread.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h [hunk 14]
```diff
@@ -60,11 +58,23 @@ class Player : public core::Thread {
...
+
     //! Stop thread.
     //! @remarks
     //!  Can be called from any thread.
     void stop();
 
+    //! Get a user defined or auto-detected sample rate of an output file or a device.
+    //!
+    //! @remarks
+    //! Output file or device should be opened.
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Two little notices:

* We can change `@remarks` to `@pre` (preconditions) here, it would fit a bit better.
* We usually indent `@remarks`, `@pre`, and similar sections, e.g.:

    ```
    //! @pre
    //!  Output file or device should be opened.
    ```

File: src/tests/roc_lib/test_sender_receiver.cpp [hunk 15]
```diff
@@ -223,6 +224,7 @@ TEST_GROUP(sender_receiver) {
         receiver_conf.n_repair_packets = n_repair_packets;
         receiver_conf.latency = packet_len * 20;
         receiver_conf.timeout = packet_len * 300;
+        receiver_conf.sample_rate = pipeline::DefaultSampleRate;
```
Comment (gavv) [inline src/tests/roc_lib/test_sender_receiver.cpp]: Since we have a default back, this line can be removed.

File: src/tests/roc_sndio/target_sox/test_player_recorder.cpp [hunk 16]
```diff
@@ -115,41 +115,47 @@ TEST_GROUP(player_recorder) {
...
     core::TempFile file("test.wav");
     CHECK(player.open(file.path()));
 
     player.stop();
-    player.start();
+    player.start(receiver);
     player.join();
 }
 
+TEST(player_recorder, player_get_sample_rate) {
```
Comment (gavv) [inline src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: I think we should test three cases:

* The user provided zero sample rate, SoX selected non-zero sample rate, Player was successfully opened.

* The user provided non-zero sample rate, SoX selected the same sample rate, Player was successfully opened.

* The user provided non-zero sample rate, but SoX selected some other sample rate. I think Player::open should fail in this case, because it's generally not a good idea to ignore what the user have asked explicitly.

I'm not sure if it's possible to add a test for the latter case because our tests operate with files instead of devices. However, we need to fix the behaviour anyway.

File: src/tools/roc_recv/main.cpp [hunk 17]
```diff
@@ -148,6 +148,22 @@ int main(int argc, char** argv) {
     core::BufferPool<audio::sample_t> sample_buffer_pool(allocator, MaxFrameSize, 1);
     packet::PacketPool packet_pool(allocator, 1);
 
+    sndio::Player player(sample_buffer_pool, allocator, args.oneshot_flag,
+                         config.channels, config.sample_rate);
```
Comment (gavv) [inline src/tools/roc_recv/main.cpp]: See https://github.com/roc-project/roc/pull/113#discussion_r145893832

File: src/tests/roc_lib/test_sender_receiver.cpp (hunk 15 above)
Comment (dshil) [reply src/tests/roc_lib/test_sender_receiver.cpp]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h (hunk 14 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Fixed

File: src/modules/roc_sndio/target_sox/roc_sndio/player.h (hunk 13 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.h]: Fixed

File: src/modules/roc_audio/resampler_updater.cpp (hunk 11 above)
Comment (dshil) [reply src/modules/roc_audio/resampler_updater.cpp]: o my gosh, at the beginning I had a crazy idea, that we can overflow here, but after your comment, I've reread the code. Both values are near about 1, we should throw this cast.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 12 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: fixed

File: src/modules/roc_pipeline/config.h (hunk 4 above)
Comment (dshil) [reply src/modules/roc_pipeline/config.h]: Fixed

File: src/tests/roc_sndio/target_sox/test_player_recorder.cpp (hunk 16 above)
Comment (dshil) [reply src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: Fixed

File: src/modules/roc_audio/resampler_updater.cpp (hunk 11 above)
Comment (gavv) [reply src/modules/roc_audio/resampler_updater.cpp]: I don't get it. If both arguments are floats, this cast simply does nothing.

File: src/modules/roc_audio/resampler_updater.h [hunk 18]
```diff
@@ -34,8 +34,11 @@ class ResamplerUpdater : public packet::IWriter,
     //! @b Parameters
     //!  - @p update_interval defines how often to call FreqEstimator, in samples
     //!  - @p aim_queue_size defines FreqEstimator target queue size, in samples
+    //!  - @p sample_rate_coef represents an adjustment coefficient for a resampler
+    //!       scaling factor, calculating from sample rates of input and output devices.
```
Comment (gavv) [inline src/modules/roc_audio/resampler_updater.h]: > from sample rates of input and output devices

Well, strictly speaking it's not rates of the input and output devices. It's a faсtor to convert session rate to receiver output rate.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp [hunk 19]
```diff
@@ -71,18 +65,49 @@ bool Player::open(const char* name, const char* type) {
         return false;
     }
 
+    size_t in_rate = size_t(out_signal_.rate);
+    size_t out_rate = size_t(output_->signal.rate);
+
+    if (in_rate != 0 && in_rate != out_rate) {
+        roc_log(LogError, "player: sample rates mismatch, auto-detected=%zu, input=%zu",
```
Comment (gavv) [inline src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: I think the error message should be something like:

"can't open output file or device with the required sample rate: required=%lu suggested=%lu".

Also some notes:

* we don't use %z, it's not portable
* we don't use commas when logging multiple values, they don't increase readability, but they make grepping a bit harder

File: src/tests/roc_lib/test_sender_receiver.cpp [hunk 20]
```diff
@@ -22,6 +22,7 @@
 #include "roc_packet/address_to_str.h"
 #include "roc_packet/packet_pool.h"
 #include "roc_packet/parse_address.h"
+#include "roc_pipeline/config.h"
```
Comment (gavv) [inline src/tests/roc_lib/test_sender_receiver.cpp]: This line can be removed I guess.

File: src/tests/roc_sndio/target_sox/test_player_recorder.cpp [hunk 21]
```diff
@@ -115,41 +115,60 @@ TEST_GROUP(player_recorder) {
...
+    Player player(buffer_pool, allocator, true, ChMask, SampleRate);
+
+    core::TempFile file("test.wav");
+    CHECK(player.open(file.path()));
+    CHECK(player.get_sample_rate() == SampleRate);
+}
+
+TEST(player_recorder, player_open_sample_rates_mistmatch) {
+    Player player(buffer_pool, allocator, true, ChMask, SampleRate + 1);
+    CHECK(!player.open(NULL, NULL));
```
Comment (gavv) [inline src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: There are three problems with this test:

* Opening an audio device may have side-effects. For example, if the user is using ALSA without dmix, only one process can open an audio device. So this test will either fail to open it (if it's already opened by someone else), or prevent other processes from opening it for a short time.

* If this test will fail to open device for some other reason (e.g. an ALSA device is already opened by someone else), it will pass anyway, so we really don't known if it actually tests something.

* I'm not sure that any SoX backend is guaranteed to fail with rate 44100+1.

We theoretically may want a separate set of tests that work with a real audio device. Such tests should not be run automatically, but instead only when the user explicitly asked that. However I'm not sure it's worth it.

Thus, if we can't cover this case with a test that uses a temporary file instead of real audio device, I suggest to remove it for now.

Comment (dshil) [reply src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: I think that you've reviewed an old version of tests, I've rebased recently.

Comment (gavv) [reply src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: The first two points are still relevant after rebase.

File: src/modules/roc_sndio/target_sox/roc_sndio/player.cpp (hunk 19 above)
Comment (dshil) [reply src/modules/roc_sndio/target_sox/roc_sndio/player.cpp]: Fixed

File: src/tests/roc_sndio/target_sox/test_player_recorder.cpp (hunk 21 above)
Comment (dshil) [reply src/tests/roc_sndio/target_sox/test_player_recorder.cpp]: Fixed

File: src/tests/roc_lib/test_sender_receiver.cpp (hunk 20 above)
Comment (dshil) [reply src/tests/roc_lib/test_sender_receiver.cpp]: Fixed

File: src/modules/roc_audio/resampler_updater.h (hunk 18 above)
Comment (dshil) [reply src/modules/roc_audio/resampler_updater.h]: Fixed
